### ISSUE TITLE
R.propOr not as useful as lodash's _.result; should it throw if object is undefined?

### ISSUE THREAD

**Issue by Andrew-Webb** [ramda/ramda#848]
This is nice in lodash:-

        var people = [
            {
                name: "Dave",
                age: 30
            },
            {
                name: "Anne",
                age: 28
            },
        ];

        _.result(_.find(people, function (person) {
            return person.age === 40;
        }), "name", undefined);

Depending on the result of the `person.age === N` test, either the name of the person or undefined will be returned.

The Ramda equivalent using propOr (is there a better way?) isn't useful:-

        R.propOr(undefined, "name", R.find(R.propEq('age', 40))(people));

This throws with "Uncaught TypeError: Cannot convert undefined or null to object".

What makes the lodash function work is the fact that _.result can cope with the result of the find being undefined (or null).  R.propOr can't cope with this.  I think that it (and similar functions) should be able to cope with undefineds and nulls - especially since the result of many list functions (returning undefined) might well be piped into them.

(Yes, I've got the first two args to R.propOr the right way round.  The documentation has it wrong; I've raised a separate issue about that.)

**Comment by CrossEye** [comment]
This is a very interesting question.  Ramda generally tries to do little hand-holding.  But throwing an exception here doesn't seem particularly useful.

I think I would need more convincing, though, that we could comfortably know the right result of

    propOr(value, name, refToNothing)

Should it be `value`?  `undefined`? `null`?  Or should it actually throw an exception?  I don't see a great argument for any of the alternatives.  Can you convince me?

**Comment by Andrew-Webb** [comment]
One of the fundamental tenets of functional programming and in Ramda is the composition of functions in a pipeline.  Therefore the behavior of a function like propOr wrt an undefined object should not be considered in isolation, but rather as part of Ramda's whole notion of function pipelining.  After all, as an API user I'm _not_ going to write this:-

    var result = R.propOr('n/a', 'name', undefined);

Since many Ramda functions can return undefined, it then follows that, for the sake of pipelining, every function that a) takes a data arg (or args) and b) might reasonably be part of a pipeline, needs to have a story in the code and in the documentation for what happens if that arg gets piped in as undefined or null.

The worst thing is to have no story.  Then the API user has to experiment to see what happens (or get bitten unexpectedly later on), and the likelihood of a good thing happening is low.  And if a good thing happens, can it be relied upon since there's no documentation?

A better thing, but still pretty bad, is to have a story and that story is: "throws an exception if the arg is specified as undefined or null".  At least we know where we are, but throwing exceptions is awful because they break the pipeline.

The best thing (in fact the only *good* thing) is to have a story and that story is aimed at not throwing an exception, but rather to aid pipelining.  So in the case of propOr, and to answer your question, the correct result of

    propOr(value, name, refToNothing)

is `value`.  This even fits the current description of the function:-

    If the given object has an own property with the specified name, returns the
    value of that property. Otherwise returns the provided default value.

Although it would be nice to augment this description with:-

    If the given object is undefined or null, the default value is returned.

I have experience of another API that supports method chaining [where thought was not given to the fact that args could well be null](http://json.codeplex.com/discussions/286048) (which caused the API to throw).  These kinds of APIs are not nice to work with.

Now, to be clear, I'm not asking for an isolated change to the behavior of propOr.  I'm talking about a considered review of each and every function in Ramda, wrt adding that nice story I talked about, in the code and in the documentation, for the data param(s) on the right of the param list.  Such a thing would really make Ramda useful and usable, and fits right in with your own stated philosophy of:-

    We try to make it clear what our functions do, to ensure there are no surprises.

**Comment by TheLudd** [comment]
@Andrew-Webb What about
```
R.compose(R.propOr('N/A', 'name'), R.defaultTo({}), getMyPerhapsUndefinedObject)
```

Btw where did you get the description of the function from?

> > If the given object has an own property with the specified name, returns the
value of that property. Otherwise returns the provided default value.

When I read the ramda docs, they say that the object should be non null:

> If the given, non-null object has an own property with the specified name, returns the value of that property. Otherwise returns the provided default value.

**Comment by Andrew-Webb** [comment]
> Btw where did you get the description of the function from?

http://ramdajs.com/ramdocs/docs/R.html#propOr

Are there two versions of the documentation, slightly differing?

**Comment by Andrew-Webb** [comment]
If I search Google for "ramda", the first result is for "ramda - npm" and the second is for "Ramda: practical functional javascript" which links to http://ramdajs.com/ramdocs/docs/.

I was using those docs.

**Comment by Andrew-Webb** [comment]
@TheLudd 

> What about
> R.compose(R.propOr('N/A', 'name'), R.defaultTo({}), getMyPerhapsUndefinedObject)

Thanks for the heads-up about defaultTo.  Being a Ramda newbie, I hadn't yet spotted that.  Coming back to my first example, I could modify it to be:-

    R.propOr(undefined, "name", R.defaultTo({}, R.find(R.propEq('age', 40))(people)));

I think it's clunky, less readable and of course more typing, but it's doable.  For a nice API usage experience, I wouldn't want to have to resort to things like that.

**Comment by CrossEye** [comment]
> If I search Google for "ramda", the first result is for "ramda - npm" and the second is for "Ramda: practical functional javascript" which links to http://ramdajs.com/ramdocs/docs/.

I thought we'd killed those ones off a while ago.  Hopefully, I just managed too with a quick hack from my phone.  Will try something more substantial later.

**Comment by CrossEye** [comment]
Also note, the documentation now lives at http://ramdajs.com/docs/

**Comment by CrossEye** [comment]
I'm not ignoring the main issue, BTW.  Just need a little more thinking.

**Comment by davidchambers** [comment]
`R.propOr` is not, in my view, well suited to pipelining. As an example, let us get the list of strings from a data structure we *hope* will have the form:

```javascript
{a: {b: {c: ['foo', 'bar', 'baz']}}}
```

As a starting point, consider the imperative solution:

```javascript
o.a == null ? [] : o.a.b == null ? [] : o.a.b.c == null ? [] : o.a.b.c
```

Here's a possible Ramda solution:

```javascript
R.pipe(R.propOr({}, 'a'),
       R.propOr({}, 'b'),
       R.propOr([], 'c'))
```

The `R.propOr` solution is okay, but it does complect each step with the step following: in the first and second steps we provide a safe intermediate value to prevent the next step from throwing.

[Sanctuary][1] tackles this issue head-on:

```javascript
R.pipe(S.get('a'),
       R.chain(S.get('b')),
       R.chain(S.get('c')),
       S.fromMaybe([]))
```

By acknowledging that certain computations may fail, Sanctuary makes it possible to compose "unsafe" functions. Assume that the uncertain data structure above is actually sent as JSON in a request body, so we must parse the JSON string before performing the property lookups.

Possible Ramda solution:

```javascript
R.pipe(function(body) {
         try {
           return JSON.parse(body);
         } catch (err) {
           return {};
         }
       },
       R.propOr({}, 'a'),
       R.propOr({}, 'b'),
       R.propOr([], 'c'))
```

Since Ramda does not acknowledge that operations may fail, it's necessary for the parse function to return a safe intermediate value.

Here's the equivalent Sanctuary code:

```javascript
R.pipe(S.parseJson,  // sugar for S.encase(JSON.parse)
       R.chain(S.get('a')),
       R.chain(S.get('b')),
       R.chain(S.get('c')),
       S.fromMaybe([]))
```

I've found Ramda + Sanctuary to be a wonderful pairing. :)


[1]: https://github.com/plaid/sanctuary

**Comment by Andrew-Webb** [comment]
FWIW, in my real code, the collection (or array, or list) I'm searching through is a cache of resources retrieved from a REST API.  Each resource object is stored along with the particular query data that was used to retrieve it.  So I search the list by query data... if the resource + query is found, I want the resource; if not I query the API and store the result (+ query data) in the cache.

From this perspective, with this use case, a find call that returns undefined (item not stored in cache) is not a 'failure'.  It's a perfectly valid result.  And, of course, the find operation shouldn't be classed as an 'unsafe' function call.

So... it would be nice if Ramda allowed me to pipe the not-in-cache result to a following function like R.propOr without recourse to R.defaultTo calls, or having a medley of R and S calls (i.e. adding Sanctuary to the mix).  Neither of those options are terrible, but they do add what I see as unwarranted complexity, and (worse) a dilution of the declarative nature of the code.  I.e. the _intent_ of my code is being somewhat hidden because I'm having to add extra jazz to stop the pipeline breaking via a thrown exception.

**Comment by davidchambers** [comment]
> From this perspective, with this use case, a find call that returns undefined (item not stored in cache) is not a 'failure'.

I disagree. `find`, like `head`, is an operation with a success case and a failure case:

    -- Ramda
    find :: (a -> Boolean) -> [a] -> a?
    head :: [a] -> a?

    -- Sanctuary
    find :: (a -> Boolean) -> [a] -> Maybe a
    head :: [a] -> Maybe a

The only difference between Ramda and Sanctuary is how success and failure are represented.

> [T]he find operation shouldn't be classed as an 'unsafe' function call.

As I see it `R.find` is unsafe. It does not guarantee a consistent return type therefore one must type check (or value check) the return value before giving it to another function. Changing *other* functions to treat `undefined` specially is not a path I'd like us to go down. How would `R.toLower(R.head([]))` evaluate? I see no option but to throw. It doesn't make sense to provide `undefined` to `R.toLower` in place of a string. Similarly it doesn't seem right to provide `undefined` to `R.propOr` in place of an object.

We could decide that `null`/`undefined` should be equivalent to `{}` throughout the API, in the same way Clojure treats `nil` as an empty collection in many contexts. This doesn't sit quite right with me, though.

**Comment by CrossEye** [comment]
I'm really not worried about the `propOr` case.  Honestly, the user has expressed her intent in the case the property is not found on the object; I think it's reasonable to extend that to include the case when the object does not exist.  The only question then would be whether the exception logic is worth it or not.

But I don't like the more general principle.  Ramda to my mind should act like a number of clearly well-defined low-level tools that don't try to guess the user's intent.

I very much like what @davidchambers is doing with Sanctuary.  I expect to be using it soon in conjunction with Ramda.  I will probably continue to object to folding it into Ramda's core, though, as being too alien to the standard Javascript ecosystem.  Most people getting the `head` of [1, 2, 3] in JS would not be happy to get back `Just(1)` when they were expecting `1`.  But as a sister library for those who want it, I think it's ideal.  And I want to ensure that Ramda is fully able to work with it.

But that is operating on a higher level.  Ramda is designed as a fairly low-level library.  It is meant to give one the primitives to work with.  We hope they're the widely useful primitives, and we're not trying to be ascetics about it.  (Yes, you can have `map` and `filter` too, even though we know you could build them out of `reduce`!)  But we try to stay well away from anything that might smack of dictating users' workflow.

In short, I would be willing to consider a not-too-invasive PR for `propOr`, but am probably not willing to use it to revamp our handling of potential `null`/`undefined` values throughout Ramda.

Oh, and @davidchambers, your simpler example could also be handled fairly easily with existing Ramda tools: `R.pipe(R.path('a.b.c'), R.defaultTo([]))`.

**Comment by Andrew-Webb** [comment]
We're used to functions that return multiple items returning an empty list/array/collection in the 'failure' case.  jQuery and C#'s LINQ (and others) show how useful this is: the next function or method down the line has something to work with.

So for functions that return a single item, returning `{}` instead of null/undefined is operating on the same principle.  The multiple-item case and the single-item case are logically equivalent.

I'm not calling on Ramda to do this, but I do think it would be interesting to have a new JS-FP library that embeds this principle at its core, from the start.

**Comment by CrossEye** [comment]
I have had minimal experience with .NET, so I can't speak to LINQ.  Overall, I think most of the jQuery API decisions are ones to be avoided.   But I do like the way jQuery wraps DOM nodes in a simple iterable collection.

But they are in a very different situation.  We rarely can provide sensible defaults for error cases.   On a `find` for instance, what would be sensible to provide in the case nothing in the list matches the predicate?   Certainly not an empty object; perhaps they expect back a string or a number.  And most functions are like that. Some do work as you'd like.   `tail` on an empty list returns an empty list. 

But the number of functions where we can provide a sensible default for error or fallback conditions seems quite small.   Obviously, `propOr` is actually one of them.

**Comment by Andrew-Webb** [comment]
> On a find for instance, what would be sensible to provide in the case nothing in the list matches the predicate?

In F# programming, the answer is known to all: you return an option.  I.e.

    Some a | None

Then, in a function that might take the result, you do pattern matching on the option, effectively combining a switch statement and a let binding (variable assignment, if you will) in one go.  Also: many standard library functions work with options in a really nice way.

When you come to consider adding some degree of pattern matching to Ramda, as you have [indicated](http://fr.umio.us/the-philosophy-of-ramda/), these sorts of possibilities (I hesitated to say "options") might present themselves.  I have a sneaking suspicion that such ideas might best be explored in a new, experimental library - for fear of breaking Ramda (which I'm enjoying using very much).

btw: one thing F# tries very hard to do, despite having to interop with standard .NET, is avoid nulls, the billion dollar mistake.

**Comment by buzzdecafe** [comment]
> you return an option

please see ramda-fantasy and plaid/sanctuary. we are interested in supporting this style of "type-safe" (in quotes because, you know, safety in JavaScript???) programming, but not in baking it in to the core.

**Comment by CrossEye** [comment]
We did talk about (#683) adding `Maybe`s to the library, but @buzzdecafe and I resisted @davidchambers' overtures.  That was the genesis of Sanctuary.  I think it makes for a good companion to Ramda for those who want to work in that manner.

But using an output like `Maybe.Nothing` as the result of `find` in your original example would not be the same as using `undefined`.  That would take quite a bit more work with Ramda to make it work properly with `propOr`.  I simply don't see that happening.
